### DISCUSSION TITLE
Double Reporting in GC

### DISCUSSION THREAD

**Discussion by SzpejnaDawid** [dotnet/runtime#108961] [Q&A] [answered]
Working on port for RISC-V I run some coreFX tests with `DOTNET_GCStress=0x4` flag and it results in the explosion of assert in `CheckDoubleReporting(...)` (path `src/coreclr/vm/gcenv.ee.common.cpp`). An object from register or stack slot sometimes is doubled. I assume that stackwalker class is correct and `REGDISPLAY` is updated correctly, so the problem is with `GCInfo`(?). But could you tell me why an object cannot occur more than once in any slot? Why does this constraint change for pinned objects?

**Comment by hez2010** [comment]
See https://github.com/dotnet/runtime/pull/107986, double reporting typical means there're potential GC holes in the implementation. 

> Double reporting is not allowed unless the reference is pinned, since it would result in incorrect updating of a reference in the slot in case the GC moves the object.

**Reply by SzpejnaDawid** [reply]
Yes yes, I saw that PR but could you tell me why double reported root means means GC hole? I understand roots (stack and register slots) come from stack walking and GCHeap::Promotion function is run for every of them. But before that there is double checking. Why cannot the same object be reported more than once?

**Reply by jkotas** [reply]
Double reporting is a problem when the GC is moving objects around. If the same object reference is reported twice when the GC is moving object around, it will be moved twice that will turn it into corrupted pointer.

Notice that the double reporting check is only done when `promotion=true` https://github.com/dotnet/runtime/pull/107986/files#diff-7f60c805631e6327d87541198da9c5b6eae87f7b0325a150ccad3e7db5a6883dR150 . `promotion=true` means that the GC is moving objects around. The double reporting check is not done for `promotion=false` when the GC just wants to know about the stack roots without updating them.

**Reply by SzpejnaDawid** [reply]
Thx @jkotas, I understand now reasons for it. But what with the situation like that:
```CSharp
void MyMagicalFun(MyClassOfObjects my_obj)
{
    MyClassOfObjects ref1 = my_obj;
    MyClassOfObjects ref2 = my_obj;
    /*
        rest of the code which uses those variables
    */
}
```
`Ryujit` can push `ref1` and `ref2` on the stack and there are at least two stack roots which point to the same object. As I understand even if the above code is correct, they cannot be reported twice when `promote = TRUE`. So... How is it done that the object is reported only once?

**Reply by jkotas** [reply]
The double reporting check is concerned about the stack root location, not the object that the stack root points to. The double reporting checks that one stack root location is not reported twice. In your example, there should be 3 stack root locations `my_obj`, `ref1`, `ref2`, each of them should have its own distinct address (assuming no optimizations).

**Reply by SzpejnaDawid** [reply]
@jkotas, Could tell me whether (in your opinion) my conclusion is correct that double reporting can only be caused by incorrect `GCInfo` generation if `stackwalking` and `REGDISPLAY` are correct? Maybe there is an aspect of GC which i didn't take into consideration.

**Reply by jkotas** [reply, marked as answer]
Yes, that's correct conclusion. The double reporting issues are typically bugs in GCInfo generation (or in stackwalking - but that's less common).

**Reply by SzpejnaDawid** [reply]
THX for conversation
